common/src/main/java/org/conscrypt/NativeSsl.java

```diff
@@ -663,13 +663,29 @@ int getPendingWrittenBytes() {
         }
 
         int writeDirectByteBuffer(long address, int length) throws IOException {
-            return NativeCrypto.ENGINE_SSL_write_BIO_direct(
-                    ssl, NativeSsl.this, bio, address, length, handshakeCallbacks);
+            lock.readLock().lock();
+            try {
+                if (isClosed()) {
```

Comment:
Should this check for `bio == 0L` ?

Reply:
Either that or make the native code throw on `bio == 0` rather than return -1 as it does at present, which doesn't seem to be handled consistently.